### PR TITLE
[Container Filter] Fix metrics exclusion for corechecks

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
Put `IsExcludedByAnnotation` inside `IsExcluded` and fix an issue where corechecks metrics were not excluded by `ad.datadoghq.com/exclude: "true"`. Corechecks use a `MetricsFilter` which previously only looked for `ad.datadoghq.com/metrics_exclude` annotations and ignored global exclusion annotations. 

### Motivation
`ad.datadoghq.com/exclude: "true"` should filter all metrics from the pod.

### Additional Notes
Linter went wild.
I can see one log not filtered out tagged with `source: kubernetes`:
```
Events from the Pod workload-nginx/nginx-555d66f9f8-bbgrf
```

### Possible Drawbacks / Trade-offs
Using `Store().GetKubernetesPodForContainer` can be quite costly. I am not sure whether we should use `Store().GetContainer(container.Owner.ID)`

### Describe how to test/QA your changes
All tests should pass.
See QA section here: https://github.com/DataDog/datadog-agent/pull/16489.
Start a pod with annotation `ad.datadoghq.com/exclude: "true"` and verify that the intake does not receive any metric tagged with the given pod_name except for `kubernetes*` metrics.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/util/containers/filter.go [hunk 1]
```diff
@@ -309,17 +311,17 @@ func NewAutodiscoveryFilter(filter FilterType) (*Filter, error) {
 		includeList = config.Datadog.GetStringSlice("container_include_logs")
 		excludeList = config.Datadog.GetStringSlice("container_exclude_logs")
 	}
-	return NewFilter(includeList, excludeList)
+	return NewFilter(ft, includeList, excludeList)
 }
 
 // IsExcluded returns a bool indicating if the container should be excluded
 // based on the filters in the containerFilter instance. Consider also using
 // IsExcludedByAnnotation in parallel with this check.
```
Comment (sblumenthal) [inline pkg/util/containers/filter.go:319]: this second sentence can be removed as a result of this change

**Review by sblumenthal — COMMENTED** [review]
one comment, love the change

File: pkg/util/containers/filter.go [hunk 2]
```diff
@@ -364,24 +366,29 @@ func (cf Filter) IsExcluded(containerName, containerImage, podNamespace string)
 		}
 	}
 
-	return false
+	return cf.isExcludedByAnnotation(annotations, containerName)
```
Comment (sblumenthal) [inline pkg/util/containers/filter.go:369]: I think we might want to move this up further in the function, mainly because i think we want annotations to be used to override the env vars, but also because the annotation check should be faster than the env checks would be, so if people are using the annotation version they would then see improved performance (or maybe not, it is so slight anyway)

File: pkg/util/containers/filter.go (hunk 1 above)
Comment (AliDatadog) [reply pkg/util/containers/filter.go:319]: Good catch ! I'll remove it

File: pkg/util/containers/filter.go (hunk 2 above)
Comment (AliDatadog) [reply pkg/util/containers/filter.go]: Good catch ! It should be fixed by now.

File: pkg/util/containers/filter.go [hunk 3]
```diff
@@ -367,21 +372,25 @@ func (cf Filter) IsExcluded(containerName, containerImage, podNamespace string)
...
-		return isExcludedByAnnotation(annotations, containerName, "metrics_")
+		return isExcludedByAnnotationInner(annotations, containerName, "metrics_")
 	case LogsFilter:
-		return isExcludedByAnnotation(annotations, containerName, "logs_")
+		return isExcludedByAnnotationInner(annotations, containerName, "logs_")
+	default:
+		log.Warnf("unrecognized filter type: %s", cf.FilterType)
 	}
-	return false
+	return isExcludedByAnnotationInner(annotations, containerName, "")
```
Comment (AliDatadog) [inline pkg/util/containers/filter.go:390]: I noticed that we should always check for `GlobalFilter`. Otherwise, the `SharedMetricFilter` (retrieved with `GetSharedMetricFilter()` will not filter out corecheck metrics.

**Review by sblumenthal — APPROVED** [review]

File: pkg/collector/python/containers.go [hunk 4]
```diff
@@ -41,7 +41,7 @@ func IsContainerExcluded(name, image, namespace *C.char) C.int {
 		goNs = C.GoString(namespace)
 	}
 
-	if filter.IsExcluded(goName, goImg, goNs) {
+	if filter.IsExcluded(nil, goName, goImg, goNs) {
```
Comment (AliDatadog) [inline pkg/collector/python/containers.go:44]: This code is only used by the `Kubelet` check. Should we also add filtering by pod annotations for Python code ?

File: pkg/collector/corechecks/containers/containerd/events.go [hunk 5]
```diff
@@ -39,7 +39,7 @@ func computeEvents(events []containerdEvent, sender aggregator.Sender, fil *cont
 			continue
 		}
 
-		if split[1] == "images" && fil.IsExcluded("", e.ID, "") {
+		if split[1] == "images" && fil.IsExcluded(nil, "", e.ID, "") {
```
Comment (AliDatadog) [inline pkg/collector/corechecks/containers/containerd/events.go:42]: I don't know if filtering by annotations should be applied here as well.

File: pkg/collector/corechecks/containers/docker/check.go [hunk 6]
```diff
@@ -211,7 +211,12 @@ func (d *DockerCheck) runDockerCustom(sender aggregator.Sender, du docker.Client
 		if len(rawContainer.Names) > 0 {
 			containerName = rawContainer.Names[0]
 		}
-		isContainerExcluded := d.containerFilter.IsExcluded(containerName, resolvedImageName, rawContainer.Labels[kubernetes.CriContainerNamespaceLabel])
+		var annotations map[string]string
+		if pod, err := workloadmeta.GetGlobalStore().GetKubernetesPodForContainer(rawContainer.ID); err == nil {
+			annotations = pod.Annotations
```
Comment (vickenty) [inline pkg/collector/corechecks/containers/docker/check.go:216]: We check annotations where we didn't before. Does this mean that there will be more things that can be excluded with the new code?

File: pkg/util/containers/filter_test.go [hunk 7]
```diff
@@ -400,16 +425,16 @@ func TestNewMetricFilterFromConfig(t *testing.T) {
 	f, err := newMetricFilterFromConfig()
 	require.NoError(t, err)
 
-	assert.True(t, f.IsExcluded("dd-152462", "dummy:latest", ""))
-	assert.False(t, f.IsExcluded("dd-152462", "apache:latest", ""))
-	assert.False(t, f.IsExcluded("dummy", "dummy", ""))
-	assert.True(t, f.IsExcluded("dummy", "k8s.gcr.io/pause-amd64:3.1", ""))
-	assert.True(t, f.IsExcluded("dummy", "rancher/pause-amd64:3.1", ""))
+	assert.True(t, f.IsExcluded(nil, "dd-152462", "dummy:latest", ""))
```
Comment (vickenty) [inline pkg/util/containers/filter_test.go:428]: Would it make sense to add tests for IsExcluded when the first argument is not nil?

Comment (AliDatadog) [reply pkg/util/containers/filter_test.go:428]: We have other tests covering cases when the first argument is not `nil`. I believe they might be enough.

File: pkg/collector/corechecks/containers/docker/check.go (hunk 6 above)
Comment (AliDatadog) [reply pkg/collector/corechecks/containers/docker/check.go]: Indeed, we can now exclude pods or containers by adding pod annotations.
This feature was added in https://github.com/DataDog/datadog-agent/pull/16489 but did not filter out corecheck metrics.

**Review by mbotarro — APPROVED** [review]
looks good for Processes

File: pkg/collector/corechecks/containers/docker/check.go (hunk 6 above)
Comment (AliDatadog) [reply pkg/collector/corechecks/containers/docker/check.go]: If a customer does not use the annotation then the behavior won't change. I don't think it would be useful to add a releasenote as the current PR is just a fix of a feature that is still not released.
There is already a releasenote for this feature: https://github.com/DataDog/datadog-agent/blob/main/releasenotes/notes/ad-exclude-annotations-0cce55729c053b9d.yaml

**Review by vickenty — APPROVED** [review]
